rat/model.py

```diff
@@ -1,272 +1,340 @@
 
-        If compile_path is not None, then write the compiled model to the given path (will only overwrite
-        existing files if the overwrite flag is true)
+class CodeExecutor(NodeWalker):
+    arguments: Dict[ast.Variable, Any]
+    parameters: Dict[str, Any]
+    left_side_of_sampling: Union[None, ast.ModelBase]
 
-        The parsed_lines argument is for creating a model from an intermediate representation -- likely
-        deprecated soon.
-        """
-        match data:
-            case pandas.DataFrame():
-                data_names = data.columns
-            case dict():
-                data_names = set()
-                for key, value in data.items():
-                    if not isinstance(key, str):
-                        raise Exception(f"Keys of dictionary form of data must be of type `str`, found {type(key)}")
-
-                    match value:
-                        case pandas.DataFrame():
-                            for column in value.columns:
-                                data_names.add(column)
-                        case _:
-                            raise Exception(f"Values of dictionary form of data must be pandas DataFrames, found {type(value)}")
-            case _:
-                raise Exception("Data must be a pandas DataFrame or a dictionary")
+    def __init__(self, arguments: Dict[ast.Variable, Any] = None, parameters: Dict[str, Any] = None):
+        self.arguments = arguments
+        self.parameters = parameters
+        self.left_side_of_sampling = None
+        super().__init__()
+
+    def walk_Statement(self, node: ast.Statement):
+        if node.op == "~":
+            self.left_side_of_sampling = node.left
+            return_value = self.walk(node.right)
+            self.left_side_of_sampling = None
+            return return_value
+        else:
+            raise ExecuteException(f"{node.op} operator not supported in BaseCodeGenerator", node)
+
+    def walk_Binary(self, node: ast.Binary):
+        left = self.walk(node.left)
+        right = self.walk(node.right)
+        return eval(f"left {node.op} right")
+
+    def walk_IfElse(self, node: ast.IfElse):
+        predicate = self.arguments[node.predicate]
+        return jax.lax.cond(predicate, lambda: self.walk(node.left), lambda: self.walk(node.right))
 
-        if model_string is not None:
-            if parsed_lines is not None:
-                raise Exception("Only one of model_string and parsed_lines can be non-None")
+    def walk_FunctionCall(self, node: ast.FunctionCall):
+        argument_list = []
 
-            parsed_lines = []
-            scanned_lines = Scanner(model_string).scan()
-            for scanned_line in scanned_lines:
-                parsed_lines.append(Parser(scanned_line, data_names, model_string).statement())
+        if self.left_side_of_sampling:
+            argument_list += [self.walk(self.left_side_of_sampling)]
```

Comment:
Why is the LHS variable's values being added?

Reply:
This is analogous to the Stan conversion of:

```
y ~ normal(mu, sigma);
```

to

```
target += normal_lpdf(y, mu, sigma);
```